### PR TITLE
Bluespace pipes

### PR DESCRIPTION
**Pull request by pubby**
Adds a new unary pipe component: bluespace pipes, which allow one to teleport gas.

The way it works is, every bluespace pipe is implicitly connected to every other bluespace pipe in existence. Pumping gas into one bluespace pipe shares it among all other bluespace pipes.

image

image

I'm submitting this as an alternative to #30391, as it addresses some of the problems discussed, such as how awful it is to repair distro once it gets bombed. With bluespace pipes, fixing distro is actually fun!

And let's be honest, connecting pipes over long distances is one of the least functional, and more importantly, least *fun* aspects of atmospherics. It takes so much time that nobody does it, and that's why you never see clever atmos projects around the station.

**Does this make regular pipes redundant?**

Hardly. Pipes are far superior at short range because you have control of what they share with. You also have better control over what gases goes where. With bluespace pipes, you only have a single pipenet, so you'll need a bunch of filters and mixers and pumps if you want multiple projects.

**How will this change the game**

Single-room floods become possible, and you can expect a higher chance of atmospherics getting repaired once broken. Expect to see autism projects make better use of atmos. Expect to see other creative uses (such as using this for cryo).

:cl: Pubby
add: Bluespace pipes to atmospherics, which create a single pipenet with all bluespace pipes in existence. 
/:cl:

### CODE REVIEW

**Comment by CPTANT** [comment]
Love it.

**Comment by calzilla1** [comment]
Cant you link 2 specific pipes?

**Comment by Onule** [comment]
Would it be possible to connect atmospherics in lavaland with the station's atmospherics?

File: code/modules/atmospherics/machinery/atmosmachinery.dm [hunk 1]
```diff
@@ -285,4 +296,5 @@ Pipelines + Other Objects -> Pipe network
 
 //Used for certain children of obj/machinery/atmospherics to not show pipe vision when mob is inside it.
 /obj/machinery/atmospherics/proc/can_see_pipes()
-	return 1
\ No newline at end of file
+	return 1
```
Comment (Cyberboss) [inline code/modules/atmospherics/machinery/atmosmachinery.dm]: TRUE

File: code/modules/atmospherics/machinery/pipes/bluespace_pipe.dm [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	icon_state = "map"
+	dir = SOUTH
+	initialize_directions = SOUTH
+	device_type = UNARY
+	can_buckle = FALSE
+
+/obj/machinery/atmospherics/pipe/bluespace/New()
+	icon_state = "pipe"
+	GLOB.bluespace_pipes += src
+	..()
```
Comment (Cyberboss) [inline code/modules/atmospherics/machinery/pipes/bluespace_pipe.dm]: Initialize and . = ..() first

File: code/modules/atmospherics/machinery/pipes/bluespace_pipe.dm [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+/obj/machinery/atmospherics/pipe/bluespace/Destroy()
+	GLOB.bluespace_pipes -= src
+	for(var/p in GLOB.bluespace_pipes)
+		var/obj/machinery/atmospherics/pipe/bluespace/P = p
+		qdel(P.parent)
+		P.parent = null
+	. = ..()
+	for(var/p in GLOB.bluespace_pipes)
```
Comment (Cyberboss) [inline code/modules/atmospherics/machinery/pipes/bluespace_pipe.dm]: return ..() last

**Comment by RandomMarine** [comment]
This'll take atmos sabotage potential to a new level.

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by XDTM** [comment]
Should be twinned pipes or use a channel, if it's global you can only have one single autism project with these

**Comment by SgtxRootbeer** [comment]
yeah, giving it a frequency or channel system would be a massive improvement.

**Comment by AuroranAI** [comment]
If you don't go with channels, make sure to take these out of the rpd's build list. Any shmuck who has time can print an rpd out of a hacked lathe and then fuck up your autism project by volume pumping the gas into space

**Comment by Ratvar** [comment]
Allow RPD to hold bluespace crystals, make those buildable with bluespace crystal charges?

**Comment by kevinz000** [comment]
imo they shouldn't be free (and in my very honest opinion they shouldn't be common either but whatever)

File: code/modules/atmospherics/machinery/pipes/bluespace_pipe.dm [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+/obj/machinery/atmospherics/pipe/bluespace/New()
+	icon_state = "pipe"
+	GLOB.bluespace_pipes += src
+	..()
+
+/obj/machinery/atmospherics/pipe/bluespace/Destroy()
+	GLOB.bluespace_pipes -= src
+	for(var/p in GLOB.bluespace_pipes)
+		var/obj/machinery/atmospherics/pipe/bluespace/P = p
+		qdel(P.parent)
```
Comment (lzimann) [inline code/modules/atmospherics/machinery/pipes/bluespace_pipe.dm]: `QDEL_NULL(P.parent)` and remove next line

**Comment by AnonymousNow** [comment]
If they could be paired with a multitool on a specific frequency each, for example, then they'd offer great tools for future mappers who aren't concerned about making their creations' atmos networks too physical and complex.

File: code/modules/atmospherics/machinery/pipes/bluespace_pipe.dm (hunk 2 above)
Comment (pubby) [reply code/modules/atmospherics/machinery/pipes/bluespace_pipe.dm]: I'll open another PR in a few minutes to change all atmos machines to use initialize.

**Comment by Xhuis** [comment]
I can think of so many great uses for this. This might make me an Atmos main again. 😍

**Comment by praisenarsie** [comment]
imo these should be VERY expensive since i expect people to highly cheat atmos with these.

**Comment by as334** [comment]
I think these would be best off being gated behind R&D or cargo or something as opposed to unlimited supply from the RPD for the reasons mentioned above. Besides, atmos doesn't really have any dependence on external departments so it would be reasonable to have them have to work with someone else for these.

**Comment by pubby** [comment]
@Onule  They do connect across z-levels so you can use this to send gas to lavaland.

**Comment by nicbn** [comment]
Balance this shit, make it require a lot of materials and science.

**Comment by pubby** [comment]
When has gating things behind science ever been a good way of balancing things? Why is balance even being discussed when nobody has put forth actual problems or test results? You lads are mad. Mad I say!

**Comment by RandomMarine** [comment]
Siphoning air from lavaland to refill the station would become practical!

**Comment by JJRcop** [comment]
Only one shared network in the entire game is kinda weird.

**Comment by Xhuis** [comment]
I feel like the limited networks are necessary for balance. Also, keep in mind that you can use gas filters to extract only the gas you want to have from the networks, so others can still use the pipes for their own needs.

**Comment by as334** [comment]
@pubby
>Actual problems.
Random people dumping plasma into your bluespace pipe setup.

**Comment by pubby** [comment]
That sounds like fun, as334. I doubt it would ever happen, but it sounds like fun when it does. Force those techies to use their filters.

**Comment by lzimann** [comment]
does this have any downside? feels like it could be really strong

**Comment by JJRcop** [comment]
The problem isn't necessarily filters, it's also that someone could just use a really strong pump and suck all the gas away.

File: code/modules/atmospherics/machinery/pipes/bluespace_pipe.dm (hunk 2 above)
Comment (pubby) [reply code/modules/atmospherics/machinery/pipes/bluespace_pipe.dm]: Too many pipenet related things broke during mapload when I tried changing to `Initialize` so I'm just gonna leave it for now if that's ok with you.

**Comment by Tokiko1** [comment]
Maybe add a range limit to the pipes and only have them share with bluespace pipes nearby within that range? You'd still be able to daisy chain networks together, but sabotage wouldn't be as easy as wrenching down another bluespace pipe anywhere.

Also, how about making these very power hungry to balance them? You'd have an incentive to either setup the engine well, or not get too lazy with the piping.

**Comment by AdamElTablawy** [comment]
Range limit is probably a good idea as to avoid a 2x1 room with a plasma canister connector and bluespace pump scenario where one asshole floods the station from a room in deep space

**Comment by FantasticFwoosh** [comment]
**If this is made out of RnD, they are just going to get exploited by ghostroles to steal air, or vice versa grief each other**

A way to configure seperate channels please but great idea

- Maybe bluespace pipes not put into a specific frequency can only direct gas in a direction (set by the player) to other bluespace ports? That'd mean you'd be able to jump ACROSS pipe networks without disturbing the existing pipe layout and deliver gases instantaneously rather than have omnipresent gas that needs to be filtered at all times.

As to say to the above point, you could have two rooms in space seperated, the airtank is in the leftmost side of the room and you're in the righthand side of the room with a vent, the airtank bluespace pipe is set to port to the nearest bluespace pipe in the eastern direction, and between both bluespace pipes is a window with a LOS and they connect and pipe over automatically with no fuss.

**Comment by Bawhoppen** [comment]
Don't just add them to atmos outright. Thought R&D wouldn't be a good answer either.

**Comment by JJRcop** [comment]
Let me throw a questionable idea into the hat: crafting

**Comment by juanototo2** [comment]
1 bluespace crystal + 1 pipe + welding tool with fuel + 10 sheets of metal = bluespace pipe? No, too cheap.

**Comment by Bawhoppen** [comment]
@juanototo2 Uses titanium or plastitanium perhaps?

**Comment by Tokiko1** [comment]
Please don't gate engineering related things behind mining/science. If you want to gate it, use power instead. Variable power output direly needs some use and RnD and mining have enough things already.

**Comment by Bawhoppen** [comment]
Mining is supposed to provide for the entire station, not JUST science. Science also in theory is then supposed to provide for the rest of the station too.

**Comment by SgtxRootbeer** [comment]
Give engineering a machine, that takes a lot of power to keep the bluespace pipes online, (wire it directly to the engine)
The more pipes, the more power it consumes

**Comment by Supermichael777** [comment]
the problem baw is that science is run by the extremely greedy.
You see them sit in the autism fort with a locker full of minerals, go through them slowly and then GO BACK TO GRAB THE NEXT LOAD far before they run out. most minerals end up in a pile at the feet of the rnd researcher. No one else has redemption machine access to nab the stuff they need and in all likely-hood they would be assaulted by the nerds for daring to "steal" "their" minerals.  Ever tried making a uranium drink?

**Review by Cyberboss — APPROVED** [review]

**Comment by Bawhoppen** [comment]
Sounds like we should have some administrative action on that then. And also other people do technically have access to the ORM (chemist, and stuff I think)

**Comment by CPTANT** [comment]
ORM should be placed inside cargo and cargo techs should have access.

Distributing resources fairly should be cargo's job.

**Comment by Cyberboss** [comment]
I'm going to overhaul the mineral refinement/distribution process soon so
let's not use it as a point against this PR

On Sep 21, 2017 4:18 PM, "CPTANT" <notifications@github.com> wrote:

ORM should be placed inside cargo and cargo techs should have access.

Distributing resources fairly should be cargo's job.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/30863#issuecomment-331269361>,
or mute the thread

**Comment by GunHog** [comment]
Prepare for a nice helping of burning plasma in your secret bases!! :smiling_imp:

**Comment by Qbopper** [comment]
lol

someone made the plasma pipe go into this little tile and be basically tried to bait people into opening it by talking about traitor gear in his bag (he wasn't an antag)

really makes you think

literally right after I finished typing this

lol2

**Comment by GunHog** [comment]
WAAAAHAHAHAHAHA! That is CLASSIC! You do not need Bluespace pipes to do this, but he prooobably ate a ban for doing that as non-antag. Still, 10/10! Stuff like this is why I love this game. (Except the non-antag part)

**Comment by Anonmare** [comment]
The taste for valids was too much to resist

**Comment by duncathan** [comment]
these are poorly implemented

**Comment by SgtxRootbeer** [comment]
you're just now noticing? especially after people cried about wanting channels over and over again...

**Comment by duncathan** [comment]
these were merged about 2 weeks before i returned i couldn't stop them

**Pull request merged**